### ISSUE TITLE
Handle DOWN in terminate_worker/3

### ISSUE THREAD

**Issue by markquezada** [bitcrowd/chromic_pdf#219]
Hi, 

We just started running into this issue in production:

> 2023-01-19T20:31:56.032465131Z 20:31:55.994 [error] Error during ChromicPDF.Browser.SessionPool.terminate_worker/3 callback:
> 2023-01-19T20:31:56.032514431Z ** (FunctionClauseError) no function clause matching in ChromicPDF.Browser.SessionPool.terminate_worker/3
> 2023-01-19T20:31:56.032520231Z     (chromic_pdf 1.6.0) lib/chromic_pdf/pdf/browser/session_pool.ex:191: ChromicPDF.Browser.SessionPool.terminate_worker(:DOWN, %{session: %{session_id: "B4192E860C6DB1C83D3A4404A257EFD0", target_id: "B3EDCFD2C0E3D3349617B0A1F59AEC0F"}, uses: 14}, %{browser: #PID<0.2847.0>, init_timeout: 5000, max_session_uses: 1000, spawn_protocol: %ChromicPDF.Protocol{steps: [call: &ChromicPDF.SpawnSession.create_browser_context/2, await: &ChromicPDF.SpawnSession.browser_context_created/2, call: &ChromicPDF.SpawnSession.create_target/2, await: &ChromicPDF.SpawnSession.target_created/2, call: &ChromicPDF.SpawnSession.attach/2, await: &ChromicPDF.SpawnSession.attached/2, call: &ChromicPDF.SpawnSession.set_user_agent/2, call: &ChromicPDF.SpawnSession.offline_mode/2, call: &ChromicPDF.SpawnSession.enable_page/2, await: &ChromicPDF.SpawnSession.page_enabled/2, call: &ChromicPDF.ResetTarget.reset_history/2, await: &ChromicPDF.ResetTarget.history_reset/2, call: &ChromicPDF.ResetTarget.blank/2, await: &ChromicPDF.ResetTarget.blanked/2, await: &ChromicPDF.ResetTarget.fsl_after_blank/2, output: &ChromicPDF.SpawnSession.output/1], state: %{__protocol__: ChromicPDF.SpawnSession, chrome_args: "--disable-dev-shm-usage", discard_stderr: false, ignore_certificate_errors: false, no_sandbox: true, offline: true, on_demand: false}}, timeout: 5000})
> 2023-01-19T20:31:56.032529731Z     (nimble_pool 0.2.6) lib/nimble_pool.ex:932: NimblePool.do_apply_worker_callback/4
> 2023-01-19T20:31:56.032532831Z     (nimble_pool 0.2.6) lib/nimble_pool.ex:867: NimblePool.maybe_terminate_worker/3
> 2023-01-19T20:31:56.032536031Z     (nimble_pool 0.2.6) lib/nimble_pool.ex:769: NimblePool.remove_worker/3
> 2023-01-19T20:31:56.032539231Z     (nimble_pool 0.2.6) lib/nimble_pool.ex:651: NimblePool.cancel_request_ref/3
> 2023-01-19T20:31:56.032542331Z     (stdlib 4.0.1) gen_server.erl:1120: :gen_server.try_dispatch/4
> 2023-01-19T20:31:56.032545331Z     (stdlib 4.0.1) gen_server.erl:1197: :gen_server.handle_msg/6
> 2023-01-19T20:31:56.032549331Z     (stdlib 4.0.1) proc_lib.erl:240: :proc_lib.init_p_do_apply/3

I'm not sure if it's related but we recently upgraded to ChromicPDF 1.6.0. I'm not sure how to diagnose from here. If it's any help, this is when trying to generate about 15 PDFs at the same time. This error is not happening locally for the same set of data, only on our production server.

**Comment by maltoe** [comment]
Hi @markquezada 

yes, I recently changed the `terminate_worker/3` to be more restrictive on the expected termination reasons. 

You already found the NimblePool docs. I interpreted the "the client link breaks" part as "the process that checked out the 'worker'" has died. Which is not the chrome process (rather, our `Browser` GenServer), but the process that called `print_to_pdf`, in other words your client. See [here](https://github.com/dashbitco/nimble_pool/blob/f786fc1dc86cdcc0b44235a57f1c89aaf79351c5/lib/nimble_pool.ex#L454) for where the NimblePool's pool process begins monitoring the client process and [here](https://github.com/dashbitco/nimble_pool/blob/f786fc1dc86cdcc0b44235a57f1c89aaf79351c5/lib/nimble_pool.ex#L545) for where it handles the `DOWN` message.

In the recent patch I categorized the `DOWN` reason as "should not happen" - as I had never seen it happen outside some weird test scenarios - and thus changed `terminate_worker/3` to not handle it. Mostly because I didn't want to silently allow this and turn a blind eye on potential problems it may point to. 

However, thinking a bit about it now: One scenario I could imagine is that, during shutdown, the VM terminates your client process **before** the pool process is terminated, so it first receives the `DOWN` for the link to the dead client process, before it receives the` shutdown` from its supervisor. It's merely a guess though. I'm by no means an expert on Erlang linked processes and supervision trees or the like.

Regardless, I don't really think this points to a deeper underlying issue anymore, and the fix is trivial, so I'll just change it again.

**Comment by maltoe** [comment]
@markquezada If you can provide more information on how often this happens and in which situations, that would be appreciated :bow:

**Comment by markquezada** [comment]
@maltoe Appreciate the quick response, as always!

For us, this appears to only happen when generating multiple PDF files in a single web request. The same set of PDFs generated one-by-one seem to work fine.

For context: We are normally generating PDFs with only 1 or 2 pages. But we have an option to generate a "bulk" PDF where we generate on the order of 10-100 of these single PDFs and then concatenate them together, all during a single web request. That's the request that is now failing.

I've been able to narrow it down to this error even on just 5 of these 1-page PDFs whereas before we were able to generate 100-200 of them in a single request without issue. If there's any more info I can provide, please let me know!

**Comment by maltoe** [comment]
Hmm, that sounds like the unhandled `DOWN` message is only the consequence of an error in your client request. Could it be that someone times out the web requests and kills their handler process or so? Keep in mind the `DOWN` error is raised in the *pool* process after the process that called `print_to_pdf` has died. Or rather, the link to it has been disconnected, but AFAIK that's usually the same thing.

But you're saying the request worked before the update to 1.6.0 and now it fails most of the time, yes? So something must have changed in between... Need to think about it, recap what we've changed. Can you check which version you were running before? Also, is it possible for you to downgrade for now?

**Comment by markquezada** [comment]
@maltoe yes, I can try downgrading.

But what's odd is that the same code running locally on my macbook does not fail. This appears to only happen on our cloud-based servers. And only after the recent upgrade.

I will check to see what configuration differences there are between the two and report back.

**Comment by markquezada** [comment]
`config.exs:`

```elixir
config :my_app, ChromicPDF,
  on_demand: false,
  offline: true,
  no_sandbox: true,
  chrome_args: "--disable-dev-shm-usage"
```

`dev.exs:`

```elixir
config :my_app, ChromicPDF, on_demand: true
```

`runtime.exs:`

```elixir
config :my_app, ChromicPDF, discard_stderr: env!("CHROMIC_PDF_DISCARD_STDERR", :boolean, true)
```

I tried setting `on_demand: false` in dev to see if I could reproduce that error. So far it continues to work on large requests. I just tried 2 with over 180+ PDFs and it worked fine locally.

**Comment by andreasknoepfle** [comment]
Just a wild guess, but could it be that you hit some resource limits in your cloud based servers, which causes the whole erlang process/node/whatever to shutdown and restart? As I said just a wild guess, since you mentioned you do not see this problem locally, where you might have more free resources (or at least no hard limit on them) and that you are only seeing this if you print multiple pages at once (which probably consumes more resources that just single prints).

**Comment by markquezada** [comment]
If we weren't able to generate these files before I'd say that's a valid theory. But this problem only started happening after we upgraded to 1.6. (As far as we know.)

It's a decently sized VM with 1.75GB of Memory and nothing else on it. I also don't see any other logs indicating a problem with the erlang VM 🤔

I'll do some more digging.

**Comment by maltoe** [comment]
@markquezada :+1: Thanks for investigating. We can also merge and release the patch to handle the `DOWN` in `terminate_worker`, though I wonder if we're only hiding an underlying issue then. Let us know if you want it in.

**Comment by markquezada** [comment]
Agree that it may be hiding an underlying issue. Definitely don't want that. (Also don't want it to affect others if it's merged into the lib.) Let me do some testing with that branch as well and I'll report back before you merge it in. Thanks again for the responsiveness!

**Comment by markquezada** [comment]
Quick update here: I tried using the updated branch that handles `:DOWN` messages but now I simply get:

```
2023-01-30T23:25:13.601232266Z ** (exit) exited in: Task.Supervised.stream(5000)
2023-01-30T23:25:13.601235266Z     ** (EXIT) time out
```

And ultimately it still fails so it appears to be some other underlying issue. I'm still investigating and will report back but it appears likely that this issue is not stemming from the change in ChromicPDF. (It's probably safe to close that PR, unmerged.)

**Comment by markquezada** [comment]
Ok, got to the bottom of this. Looks like this was an issue I introduced by trying to parallelize PDF rendering for a bulk PDF archive I was creating. It just happened to coincide with upgrading to 1.6 but that was a red herring.

For reference, the code in question:

```elixir
  def render(transcripts, filename) do
    transcripts
    |> Task.async_stream(&render_transcript/1, max_concurrency: 25) # <- the culprit
    |> Enum.map(fn {:ok, filepath} -> filepath end)
    |> concatenate_and_cleanup(filename)
  end
``` 

... where `render_transcript/1` is executing `ChromicPDF.print_to_pdf`. 

The default `:timeout` for `Task.async_stream` is `5000`. My guess is that the ChromicPDF job was taking more than 5s in some cases due to waiting on the pool, etc., and that timeout was getting hit, killing the processes and triggering the `:DOWN` message to `ChromicPDF.Browser.SessionPool.terminate_worker/3`.

After increasing the timeout on the async call, I got this error which seems to corroborate the theory:

```
Elixir.ChromicPDF.Browser.ExecutionError Caught EXIT signal from NimblePool.checkout!/4

      ** (EXIT) time out

This means that your operation was unable to acquire a worker from the pool
within 5000ms, as all workers are currently occupied.

Two scenarios where this may happen:

1) You suffer from this error at boot time or in your CI. For instance,
   you're running PDF printing tests in CI and occasionally the first of these test
   fails. This may be caused by Chrome being delayed by initialization tasks when
   it is first launched.

   See ChromicPDF.warm_up/1 for a possible mitigation.

2) You're experiencing this error randomly under load. This would indicate that
   the number of concurrent print jobs exceeds the total number of workers in
   the pool, so that all workers are occupied.

   To fix this, you need to increase your resources, e.g. by increasing the number
   of workers with the `session_pool: [size: ...]` option.

   However, please be aware that while ChromicPDF (by virtue of the underlying
   NimblePool worker pool) does perform simple queueing of worker checkouts,
   it is not suitable as a proper job queue. If you expect to peaks in your load
   leading to a high level of concurrent use of your PDF printing component,
   a job queue like Oban will provide a better experience.

Please also consult the worker pool section in the documentation.
 
    lib/chromic_pdf/pdf/browser/session_pool.ex:70 ChromicPDF.Browser.SessionPool.run_protocol/3
    lib/chromic_pdf/api.ex:67 anonymous fn/3 in ChromicPDF.API.chrome_export/4
    lib/chromic_pdf/api/telemetry.ex:10 anonymous fn/2 in ChromicPDF.Telemetry.with_telemetry/3
    /app/deps/telemetry/src/telemetry.erl:321 :telemetry.span/3
    lib/novo/transcript_pdf_renderer.ex:26 Novo.TranscriptPDFRenderer.render_to_file/2
    lib/novo/bulk_transcript_pdf_renderer.ex:37 Novo.BulkTranscriptPDFRenderer.render_transcript/1
    lib/task/supervised.ex:89 Task.Supervised.invoke_mfa/2
    lib/task/supervised.ex:34 Task.Supervised.reply/4
```

When I increased the session pool to 10, I got this error:

```
Elixir.ChromicPDF.Browser.ExecutionError Timeout in Channel.run_protocol/3!

The underlying GenServer.call/3 exited with a timeout. This happens when the browser was
not able to complete the current operation (= PDF print job) within the configured
5000 milliseconds.
```

So for some reason it is taking quite a while for some of these 1-page PDFs to render and that's cascading into failures as I try to process more.

Unfortunately I can't find a combination that seems to work consistently on our servers. I have now replaced the `Task.async_stream` usage with a simple `Enum.map`. It is slower but appears to have stopped the crashing.

If I'm generating the PDFs serially anyway, I suppose I can switch to the new native multi-page generation option in ChromicPDF via ghostscript but the reason I used `async_stream` in the first place was that generating a 150+ 1-page PDFs serially takes quite a while. Even a little concurrency speeds it up dramatically. That said, although the parallelized version works well on a desktop machine, our cloud servers are crashing with even a modest `max_concurrency` of `5`. (That's down from `25` I had set originally.)

I thought about doing the work to wrap a pool around our bulk PDF generator to ensure we don't overload ChromicPDF but again, I couldn't find a combination that worked with `async_stream`'s `max_concurrency` setting so I wasn't sure it was worth trying. I tried setting `max_concurrency` lower than the session pool but still experienced crashes. I don't understand why/how that could be since in theory, there should always be more than enough available in that case.

Thoughts? Is there a way to parallelize the rendering of a bunch of PDFs in a single request? Am I missing something obvious?

Side note: I likely didn't run into this in development at first because I had `on_demand: true` set in dev. When I mirrored our production setup locally, I could trigger the same errors, albeit with much higher resource limits.

**Comment by maltoe** [comment]
> Any thoughts on what might be causing the execution error or how I might proceed with debugging?

Just recapping to my sure I understood: You reverted to `Enum.map` to sequentially produce 70 1-page PDFs and some of them still times out? And there is no other job accessing ChromicPDF at the same time? How's the system load? Could it be that Chrome sometimes has to wait for resource allocation or so? I could totally imagine random delays when you're printing concurrently, e.g. due to the connection multiplexing and spikes in protocol load, but sequentially it should in theory run smoothly, no?

---

Just ran a little experiment on my machine in a project of ours where we also generate 1-page documents (not a lot of text, but some layout and styling), and the PDF generation was extremely consistent around 250ms with very small deviation. So as far as I can tell something else has to be going on in your server.

<details>
  <summary>benchmark code used</summary>
  <pre>
    def benchpdf do
    samples =
      for i <- 1..1000 do
        {uSecs, _} = :timer.tc(__MODULE__, :genpdf, [])
        ms = trunc(uSecs / 1000.0)
        IO.puts("#{ms}ms")
        ms
      end

    sum = Enum.sum(samples)

    IO.puts("Min: #{Enum.min(samples)}ms")
    IO.puts("Max: #{Enum.max(samples)}ms")
    IO.puts("Avg: #{sum / length(samples)}ms")
  end
  </pre>
</details>

---

As a solution to your problem - in case you can't figure out what's causing the timeouts - would it be an option to go fully async with a job queue (like [Oban](https://github.com/sorentwo/oban))? This would of course make synchronization in your request a lot harder, but at least then you get retries and your system should become a lot more stable overall? Just a thought.

**Comment by maltoe** [comment]
Moving forward with the issue of the `DOWN` message: 

> My guess is that the ChromicPDF job was taking more than 5s in some cases due to waiting on the pool, etc., and that timeout was getting hit, killing the processes and triggering the :DOWN message to ChromicPDF.Browser.SessionPool.terminate_worker/3.

That makes a lot of sense. I'd like to verify this on my machine before moving on with the PR on the `terminate_worker/3` callback.

**Comment by maltoe** [comment]
reproduced

<details>
  <summary>
  code
  </summary>
<pre>
Mix.install([{:chromic_pdf, path: "."}])

ChromicPDF.start_link()

# no link to this process because it is killed when `Task.async_stream/3` explodes
Process.spawn(fn ->
  long_content = String.duplicate("foobar ", 1_000_000)

  [{:ok, _}] =
    Task.async_stream([{:html, long_content}], fn source -> ChromicPDF.print_to_pdf(source) end)
    |> Enum.to_list()
end, [])

:timer.sleep(10_000)
IO.puts("DONE")
</pre>
</details>

**Comment by markquezada** [comment]
> Just recapping to my sure I understood: You reverted to `Enum.map` to sequentially produce 70 1-page PDFs and some of them still times out? 

That is correct. I wanted to eliminate the possibility of it being related to the processes, etc. And it does seem to work more consistently but it still fails with a larger number of individual PDFs. I just tried a set of 47 files and it worked. Tried 93 and it failed with:

```
2023-02-01T09:21:25.391376730Z 09:21:25.390 [error] Error during ChromicPDF.Browser.SessionPool.terminate_worker/3 callback:

2023-02-01T09:21:25.391434431Z ** (FunctionClauseError) no function clause matching in ChromicPDF.Browser.SessionPool.terminate_worker/3

2023-02-01T09:21:25.398552325Z     (chromic_pdf 1.6.0) lib/chromic_pdf/pdf/browser/session_pool.ex:191: ChromicPDF.Browser.SessionPool.terminate_worker(:DOWN, %{session: %{session_id: "5BAE11665BB7151D3E769D54C8DB45D9", target_id: "2C695FBCD9424A32EFB159176AAF8D58"}, uses: 61}, %{browser: #PID<0.2893.0>, init_timeout: 5000, max_session_uses: 1000, spawn_protocol: %ChromicPDF.Protocol{steps: [call: &ChromicPDF.SpawnSession.create_browser_context/2, await: &ChromicPDF.SpawnSession.browser_context_created/2, call: &ChromicPDF.SpawnSession.create_target/2, await: &ChromicPDF.SpawnSession.target_created/2, call: &ChromicPDF.SpawnSession.attach/2, await: &ChromicPDF.SpawnSession.attached/2, call: &ChromicPDF.SpawnSession.set_user_agent/2, call: &ChromicPDF.SpawnSession.offline_mode/2, call: &ChromicPDF.SpawnSession.enable_page/2, await: &ChromicPDF.SpawnSession.page_enabled/2, call: &ChromicPDF.ResetTarget.reset_history/2, await: &ChromicPDF.ResetTarget.history_reset/2, call: &ChromicPDF.ResetTarget.blank/2, await: &ChromicPDF.ResetTarget.blanked/2, await: &ChromicPDF.ResetTarget.fsl_after_blank/2, output: &ChromicPDF.SpawnSession.output/1], state: %{__protocol__: ChromicPDF.SpawnSession, chrome_args: "--disable-dev-shm-usage", discard_stderr: false, ignore_certificate_errors: false, no_sandbox: true, offline: true, on_demand: false}}, timeout: 5000})

2023-02-01T09:21:25.398572425Z     (nimble_pool 0.2.6) lib/nimble_pool.ex:932: NimblePool.do_apply_worker_callback/4

2023-02-01T09:21:25.398576425Z     (nimble_pool 0.2.6) lib/nimble_pool.ex:867: NimblePool.maybe_terminate_worker/3

2023-02-01T09:21:25.398587625Z     (nimble_pool 0.2.6) lib/nimble_pool.ex:769: NimblePool.remove_worker/3

2023-02-01T09:21:25.398591225Z     (nimble_pool 0.2.6) lib/nimble_pool.ex:651: NimblePool.cancel_request_ref/3

2023-02-01T09:21:25.398594625Z     (stdlib 4.0.1) gen_server.erl:1120: :gen_server.try_dispatch/4

2023-02-01T09:21:25.398598325Z     (stdlib 4.0.1) gen_server.erl:1197: :gen_server.handle_msg/6

2023-02-01T09:21:25.398605625Z     (stdlib 4.0.1) proc_lib.erl:240: :proc_lib.init_p_do_apply/3
```

> And there is no other job accessing ChromicPDF at the same time?

Correct. This is a staging server that only a few people have access to. Currently, I'm the only one online.

> How's the system load? Could it be that Chrome sometimes has to wait for resource allocation or so?

It's docker image deployed to an App Service in Azure. Theoretically shouldn't be anything else running but it could be something with shared resources if it's a VM on a larger box. I'll have to figure out if I can ssh in somehow and monitor. I will report back.

> I could totally imagine random delays when you're printing concurrently, e.g. due to the connection multiplexing and spikes in protocol load, but sequentially it should in theory run smoothly, no?

Exactly. That's what I thought. I figured moving to `Enum.map` would just make it process serially and work.

> So as far as I can tell something else has to be going on in your server.

I'll try monitoring the server itself next, to see if something else is happening. It does happen on a larger dedicated instance as well so my gut says it's something I'm doing or chrome/chromic is doing.

> As a solution to your problem - in case you can't figure out what's causing the timeouts - would it be an option to go fully async with a job queue (like [Oban](https://github.com/sorentwo/oban))? This would of course make synchronization in your request a lot harder, but at least then you get retries and your system should become a lot more stable overall? Just a thought.

Definitely possible but I'd hope to not have to resort to bringing in such a huge dependency to generate a single PDF. 🤔

**Comment by maltoe** [comment]
> I just tried a set of 47 files and it worked. Tried 93 and it failed with

Wait, but that is the same `terminate_worker` callback error that you get with `Task.async_stream`, which means the outer process has died somehow? Is this an HTTP request? Could it have been killed by its HTTP handler? Or a load balancer?

If possible, could you try that on an iex shell in the container?

> Definitely possible but I'd hope to not have to resort to bringing in such a huge dependency to generate a single PDF.

Fair point. Maybe `try...rescue` with a retry mechanism would be enough.

**Comment by markquezada** [comment]
> Wait, but that is the same `terminate_worker` callback error that you get with `Task.async_stream`, which means the outer process has died somehow? Is this an HTTP request? Could it have been killed by its HTTP handler? Or a load balancer?

Just realized that was strange since before I was getting an ExecutionError. All I can think is that we're hitting a timeout in either the web server or on Azure.

I did a quick test:

```
2023-02-01T10:36:41.558039846Z 10:36:41.556 request_id=Fz-s_-qN6GiyfEEAABsx [info] GET /transcripts/bulk-download
2023-02-01T10:37:41.571587993Z 10:37:41.570 [error] Error during ChromicPDF.Browser.SessionPool.terminate_worker/3 callback:
```

Note the timestamps: it fails after 60 seconds exactly. Obviously not a coincidence.

Doing some digging... Looks like [the default timeout for cowboy 2 is 60 seconds](https://github.com/ninenines/cowboy/blob/6cc3b0ccca6c06592c2a1881cae39945de805747/doc/src/manual/cowboy_http.asciidoc) 😩

Ok, so that would explain that error. It's getting terminated by cowboy. Ugh. Sorry for the wild goose chase.

Now I'm not sure if what I was seeing was due to the timeout all along or partly due to the concurrency issue. (Or both?) The `ExecutionError` seems like it could be a different issue. 

I'm going to update the `idle_timeout` and re-test with the original code and go from there. I will report back tomorrow.

**Comment by maltoe** [comment]
> Ok, so that would explain that error. It's getting terminated by cowboy. Ugh. Sorry for the wild goose chase.

No worries, glad we could figure at least that one out :)

> Unless you think its reasonable to be able to generate 150ish PDFs in sub 10s or so. Doesn't appear to be

Depends, but likely no. The document I've tried earlier was really simple in design and printing took around 250ms on average, you're more in the 30s - 60s range :/ Sorry.

**Comment by markquezada** [comment]
Following up here for posterity: This issue ended up being a myriad of small things coalescing into various failures on our staging server. Here are some notes for anyone who might run into a similar issue:

- Our VM instance was too small and had only one core. This was causing individual PDF generation to take between 1-5s each. If a single PDF took longer than 5s, `Task. async_stream` would kill the process since it's default timeout is 5s. I've bumped up our instance size to ensure at least two cores in our staging environments.
- Even if the PDFs did generate without error, Cowboy's timeout limit was 60s so the process would get killed if the PDFs took longer than that to generate.
- Task.async_stream works great to parallelize PDF generation (if tuned correctly)
- I was using too many processes to process the PDFs. I ended setting `Task. async_stream`'s `max_concurrency` to use the same number of sessions that Chromic uses. Something like this:


```elixir
  defp concurrency() do
    get_in(Application.get_env(:my_app, ChromicPDF), [:session_pool, :size]) ||
      max(div(System.schedulers_online(), 2), 1)
  end
```

Additionally, I was originally not seeing these errors locally in development because I had set `on_demand: true` instead of mirroring our production setup.

So far, after the above changes, we've been able to generate a large number of PDFs in a single request without issue.

Thanks again @maltoe for the willingness to help me figure this out along the way. 🙏
